cmd/gardenlet/app/bootstrap.go

```diff
@@ -0,0 +1,142 @@
+	kutil "github.com/gardener/gardener/pkg/utils/kubernetes"
+
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	certificatesv1beta1client "k8s.io/client-go/kubernetes/typed/certificates/v1beta1"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/client-go/util/keyutil"
+	componentbaseconfig "k8s.io/component-base/config"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+func bootstrapKubeconfig(
+	ctx context.Context,
+	logger *logrus.Logger,
+	gardenClientConnection config.GardenClientConnection,
+	seedClientConnection componentbaseconfig.ClientConnectionConfiguration,
+	seedConfig *config.SeedConfig,
+) ([]byte, error) {
+	if gardenClientConnection.KubeconfigSecret == nil {
+		logger.Info("No kubeconfig secret given in `.gardenClientConnection` configuration - skipping kubeconfig bootstrap process.")
+		return nil, nil
+	}
+
+	seedRESTCfg, err := utils.RESTConfigFromClientConnectionConfiguration(seedClientConnection, nil)
+	if err != nil {
+		return nil, err
+	}
+	k8sSeedClient, err := kubernetes.NewWithConfig(kubernetes.WithRESTConfig(seedRESTCfg))
+	if err != nil {
+		return nil, err
+	}
+
+	kubeconfigSecret := &corev1.Secret{}
+	if err := k8sSeedClient.Client().Get(ctx, kutil.Key(gardenClientConnection.KubeconfigSecret.Namespace, gardenClientConnection.KubeconfigSecret.Name), kubeconfigSecret); client.IgnoreNotFound(err) != nil {
+		return nil, err
+	}
+
+	if len(kubeconfigSecret.Data[kubernetes.KubeConfig]) > 0 {
+		logger.Info("Found kubeconfig generated from bootstrap process. Using it")
+		return kubeconfigSecret.Data[kubernetes.KubeConfig], nil
+	}
+
+	// kubeconfig secret not found or empty - trigger bootstrap process
+	if gardenClientConnection.BootstrapKubeconfig == nil {
+		return nil, fmt.Errorf("cannot trigger kubeconfig bootstrap process because `.gardenClientConnection.bootstrapKubeconfig` is not set")
+	}
+
+	logger.Info("No kubeconfig for garden cluster found, but bootstrap kubeconfig was given.")
+
+	// check if we got a kubeconfig for bootstrapping
+	bootstrapKubeconfigSecret := &corev1.Secret{}
+	if err := k8sSeedClient.Client().Get(ctx, kutil.Key(gardenClientConnection.BootstrapKubeconfig.Namespace, gardenClientConnection.BootstrapKubeconfig.Name), bootstrapKubeconfigSecret); err != nil {
+		return nil, err
+	}
+	if len(bootstrapKubeconfigSecret.Data[kubernetes.KubeConfig]) == 0 {
+		return nil, fmt.Errorf("bootstrap kubeconfig secret does not contain a kubeconfig")
+	}
+
+	// create certificate client with bootstrap kubeconfig in order to create CSR
+	bootstrapClientConfig, err := clientcmd.NewClientConfigFromBytes(bootstrapKubeconfigSecret.Data[kubernetes.KubeConfig])
+	if err != nil {
+		return nil, err
+	}
+	bootstrapConfig, err := bootstrapClientConfig.ClientConfig()
+	if err != nil {
+		return nil, err
+	}
+	bootstrapClient, err := certificatesv1beta1client.NewForConfig(bootstrapConfig)
```

Comment:
I thought about the same in the beginning, though, it's not possible, unfortunately, because we want to call the `approve` sub resource, and sub resources are not yet supported: https://github.com/kubernetes-sigs/controller-runtime/issues/172